src/apps/service/apps/jobs/submissions/async.clj

```diff
@@ -0,0 +1,118 @@
+  [user paths]
+  (into {} (map (juxt identity (partial get-path-list-contents user)) paths)))
+
+(defn- map-slice
+  [m n]
+  (->> (map (fn [[k v]] (vector k (nth v n))) m)
+       (into {})
+       (remove-nil-values)))
+
+(defn- map-slices
+  [m]
+  (let [max-count (apply max (map (comp count val) m))]
+    (mapv (partial map-slice m) (range max-count))))
+
+
+(defn- substitute-param-values
+  [path-map config]
+  (->> (map (fn [[k v]] (vector k (get path-map v v))) config)
+       (into {})))
+
+(defn- format-submission-in-batch
+  [submission job-number path-map]
+  (let [job-suffix (str "analysis-" (inc job-number))]
+    (assoc (update-in submission [:config] (partial substitute-param-values path-map))
+      :group      (config/jex-batch-group-name)
+      :name       (str (:name submission) "-" job-suffix)
+      :output_dir (ft/path-join (:output_dir submission) job-suffix))))
+
+(defn- submit-job-in-batch
+  [apps-client user submission job-number path-map]
+  (try+
+    (submit/submit-and-register-private-job apps-client
+                                            user
+                                            (format-submission-in-batch submission job-number path-map))
+    (catch Object _
+      (log/error (:throwable &throw-context)
+                 "batch job submission failed.")
+      {:status jp/failed-status})))
+
+(defn- preprocess-batch-submission
+  [submission output-dir parent-id]
+  (assoc submission
+    :output_dir           output-dir
+    :parent_id            parent-id
+    :create_output_subdir false))
+
+(defn submit-batch-jobs
+  "Asynchronously submits sub-jobs for a parent batch job,
+   where `ht-paths` is a list of paths of the HT Path List input files,
+   `submission` is the original parent submission map,
+   `output-dir` is the path to the output folder created for the parent job,
+   and `parent-id` is the parent job's ID to use for each sub-job's `parent_id` field.
+   A notification will be sent to the user on success or on failure to submit all sub-jobs."
+  [apps-client user ht-paths submission output-dir parent-id]
+  (future
+    (try+
+      (let [path-lists    (validate-path-lists (get-path-list-contents-map user ht-paths))
+            path-maps     (map-slices path-lists)
+            submission    (preprocess-batch-submission submission output-dir parent-id)
+            job-stats     (->> path-maps
+                               (map-indexed (partial submit-job-in-batch apps-client user submission))
+                               doall)
+            success-count (->> job-stats
+                               (filter (comp (partial = jp/submitted-status) :status))
+                               count)]
+        (if (> success-count 0)
+          (notifications/send-batch-submission-completed user
+                                                         (job-listings/list-job apps-client parent-id)
+                                                         success-count)
```

Comment:
I think it might be nice if this notification could specify not just how many succeeded, but also how many failed. Then just looking at the notification it's easy to tell if you can pretty much just trust it to run, or if you should go investigate why some of them failed. Not sure how hard that'd be to implement or if that's better done with a later implementation, but I thought I'd throw it out there.